bpo-33671: efficient zero-copy for shutil.copy* functions (Linux, OSX and Win)

giampaolo opened the pull request:
<!-- issue-number: bpo-33671 -->
https://bugs.python.org/issue33671
<!-- /issue-number -->

giampaolo committed:
have shutil.copyfileobj use sendfile() if possible

giampaolo committed:
refactoring: use ctx manager

giampaolo committed:
add test with non-regular file obj

giampaolo committed:
emulate case where file size can't be determined

giampaolo committed:
reference _copyfileobj_sendfile directly

giampaolo committed:
add test for offset() at certain position

giampaolo committed:
add test for empty file

giampaolo committed:
add test for non regular file dst

giampaolo committed:
small refactoring

giampaolo committed:
leave copyfileobj() alone in order to not introduce any incompatibility

giampaolo committed:
minor refactoring

giampaolo committed:
remove old test

giampaolo committed:
update docstring

giampaolo committed:
update docstring; rename exception class

giampaolo committed:
detect platforms which only support file to socket zero copy

giampaolo committed:
don't run test on platforms where file-to-file zero copy is not supported

giampaolo committed:
use tempfiles

giampaolo committed:
reset verbosity

giampaolo committed:
add test for smaller chunks

giampaolo committed:
add big file size test

giampaolo committed:
add comment

giampaolo committed:
update doc

giampaolo committed:
update whatsnew doc

giampaolo committed:
update doc

giampaolo committed:
catch Exception

giampaolo committed:
remove unused import

giampaolo committed:
add test case for error on second sendfile() call

giampaolo committed:
turn docstring into comment

giampaolo committed:
add one more test

giampaolo committed:
update comment

giampaolo committed:
add Misc/NEWS entry

giampaolo committed:
get rid of COPY_BUFSIZE; it belongs to another PR

giampaolo committed:
update doc

giampaolo committed:
expose posix._fcopyfile() for OSX

giampaolo committed:
Merge branch 'master' into shutil-osx-copyfile

giampaolo committed:
merge from linux branch

giampaolo committed:
merge from linux branch

giampaolo committed:
expose fcopyfile

giampaolo committed:
arg clinic for the win implementation

giampaolo committed:
convert path type to path_t

giampaolo committed:
expose CopyFileW

giampaolo committed:
fix windows tests

giampaolo committed:
release GIL

giampaolo committed:
minor refactoring

giampaolo committed:
update doc

giampaolo committed:
update comment

giampaolo committed:
update docstrings

giampaolo committed:
rename functions

giampaolo committed:
rename test classes

giampaolo committed:
update doc

giampaolo committed:
update doc

giampaolo committed:
update docstrings and comments

scoder commented on Modules/clinic/posixmodule.c.h:
```diff
@@ -3853,6 +3853,77 @@ os_write(PyObject *module, PyObject *const *args, Py_ssize_t nargs)
+
+#endif /* defined(__APPLE__) */
+
+#if (defined MS_WINDOWS)
```
Misplaced parenthesis? Even if this happens to work (oh the wonders of C!), I can't see a reason to spell it this way.

giampaolo committed:
avoid do import nt|posix modules if unnecessary

zooba reviewed (approved):
Looks good to me

zooba commented on Lib/shutil.py:
```diff
@@ -42,6 +42,14 @@
+    import nt
+
+_HAS_SENDFILE = hasattr(os, "sendfile")
+_HAS_FCOPYFILE = hasattr(posix, "_fcopyfile")
```
Spotted this just then - `posix` is not defined on Windows at this point (this is why the build failed)

giampaolo replied:
fixed, thanks

giampaolo committed:
set nt|posix modules to None if not available

giampaolo committed:
micro speedup

giampaolo committed:
update description

giampaolo committed:
add doc note

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -51,7 +51,8 @@ Directory and files operations
    Copy the contents (no metadata) of the file named *src* to a file named
-   *dst* and return *dst*.  *src* and *dst* are path names given as strings.
+   *dst* and return *dst*.  *src* and *dst* are path names given as strings
+   in the most efficient way possible.
```
This new doc snippet looks like it should be in the first sentence of this paragraph, not the second - it's the file copy that is performed in the most efficient way, not passing the path names in as strings.

giampaolo replied:
fixed

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -65,6 +66,13 @@ Directory and files operations
    file *src* points to.
 
+   .. note::
+      Internally platform-specific zero-copy syscalls are used on Linux, OSX
```
With the length of this note, and the number of functions it applies to, I'm thinking it may make sense to add a new subsection before the `copytree` and `rmtree` examples called "Platform-dependent efficient copy operations", that provides some more details on the underlying operations used, and the fallback to `copyfileobj`.

Then the various `versionchanged` notes can just link to that new subsection.

This will also give room to explain that the optimisation doesn't apply to all platforms that offer `os.sendfile`, it's specifically limited to those that allow the destination of `sendfile` to be a regular file descriptor (not those that require it to be an open socket stream).

giampaolo replied:
Good idea. Done.

ncoghlan commented on Lib/shutil.py:
```diff
@@ -81,6 +94,104 @@ def copyfileobj(fsrc, fdst, length=16*1024):
+        try:
+            sent = os.sendfile(outfd, infd, offset, blocksize)
+        except OSError as err:
+            if err.errno == errno.ENOTSOCK:
```
This path will also be hit consistently on FreeBSD and other systems that offer `os.sendfile`, but only with regular socket targets.

To avoid a performance regression on such platforms, it would be strongly preferred for there to be a build time configure check that told `shutil` that it shouldn't even try the optimised fast path, and should instead always go directly to `copyfileobj`.

Rather than a new flag in the OS module, this support could be indicated to the `shutil` module by exposing an `os._sendfile_copy` alias for `os.sendfile` only on platforms where `sendfile` supports regular file targets.

giampaolo replied:
> This path will also be hit consistently on FreeBSD and other systems that offer os.sendfile, but only with regular socket targets.

This would only be hit once though, as per `_HAS_SENDFILE = False` which gets globally set after first failed call.

ncoghlan replied:
How about in addition to setting the flag, we also do `_copyfileobj2 = copyfileobj`, to switch off the doomed-to-fail fastpath completely? That way after the first failed attempt, further attempts wouldn't even check the flag, since the "optimised" version would just be an alias for the regular implementation.

giampaolo replied:
To be honest I would prefer not to change module objects at runtime like this, even if the function is private. It makes testing and mocking more difficult (because tests rely on the private functions) and the speedup of avoiding a global `if VAR:` runtime check is negligible considering that the copying operations are strictly IO bound. To say one, we would get more benefit by globalizing the various `hasattr` checks which are present in the module and still it's unlikely we'll get any noticeable speedup.

ncoghlan commented on Lib/shutil.py:
```diff
@@ -81,6 +94,104 @@ def copyfileobj(fsrc, fdst, length=16*1024):
+                break  # EOF
+            offset += sent
+
+def _copyfileobj2(fsrc, fdst):
```
I'd suggest `_copyfileobj_fast` as the name for this.

giampaolo replied:
How about `_zerocopy_fileobj` in order to remain consistent with `_zerocopy_osx`, `_zerocopy_win` and `_zerocopy_sendfile`?

ncoghlan replied:
Given @eryksun's comments, a more technically accurate set of names might be `_fastcopy_osx`, `_fastcopy_win`, `_fastcopy_sendfile`, `_fastcopy_fileobj`, and `_GiveUpOnFastCopy`.

giampaolo replied:
Good idea. Done.

ncoghlan commented on Lib/shutil.py:
```diff
@@ -81,6 +94,104 @@ def copyfileobj(fsrc, fdst, length=16*1024):
+    #   GzipFile (which decompresses data), HTTPResponse (which decodes
+    #   chunks).
+    # - possibly others
+    if _HAS_SENDFILE:
```
Given a configure check for `_HAS_SENDFILE`, the initial check for both it and `_HAS_FCOPYFILE` can be moved up to the module level, and define `_copyfileobj_fast` differently based on the results. If both flags are false, then `_copyfileobj_fast` can be aliased directly to `copyfileobj`. (I believe Mac OS X defines both operations, so it will be important to give `_HAS_FCOPYFILE` precedence when deciding which implementation of `_copyfileobj_fast` to use)

The dynamic check to handle the case where a Python built on a newer distro is run against an older kernel (and hence still gets `ENOTSOCK` at runtime) can then also be handled by setting `_copyfileobj_fast = copyfileobj` rather than by setting a flag that needs to be checked on every call.

giampaolo replied:
See my comment above.

ncoghlan commented on Lib/test/test_shutil.py:
```diff
@@ -84,6 +111,37 @@ def rlistdir(path):
             res.append(name)
     return res
 
+def supports_file2file_sendfile():
```
Turning this into a build time configure check is essentially a matter of converting it to a small C program, and then adding it to configure.ac.

https://github.com/python/cpython/blob/master/configure.ac#L4414 is a reasonable example of a comparable check.

giampaolo replied:
I see your point. This is not exactly easy though as sendfile() is quite messy.
Both its signature and the inclusion of the right header files is platform dependent. See: https://github.com/giampaolo/pysendfile/blob/master/sendfilemodule.c
My concern is that we would end up duplicating a lot of logic from Modules/posixmodule.c

ncoghlan replied:
Yeah, given that we need the runtime check no matter what (thanks to the Python-built-against-a-newer-kernel-running-on-an-older-kernel problem), deferring any additional build time checks to a follow up patch makes sense (it also means we can be more confident that the runtime check is robust, since the FreeBSD buildbots will check it).

And given the logic tweak to swap completely back to the Python level implementation when we get `ENOTSOCK`, then the runtime overhead should be negligible (and if it turns out that it isn't, that's when we can reconsider the configure time check option).

ncoghlan reviewed (commented):
This is very cool, thank you!

My main concern would be with platforms which don't support non-socket targets in `sendfile` at all, and hence will always run through the "Try it, get ENOTSOCK, adjust things so they never try it again" sequence at runtime, even though it would be possible to make that determination at build time.

That's not necessarily a merge blocker (since we need the dynamic runtime check to handle older Linux kernels regardless), but it wouldn't certainly be a nicer way to handle those cases.

More detailed comments in line on how such a configure check might work, as well as some suggested adjustments to the documentation to avoid having to repeat the same explanatory snippet several times.

giampaolo committed:
use better wording in doc

eryksun commented on Modules/posixmodule.c:
```diff
@@ -8745,6 +8749,62 @@ posix_sendfile(PyObject *self, PyObject *args, PyObject *kwdict)
+#endif
+
+
+#if defined MS_WINDOWS
```
This is not a POSIX function, so I think it should be implemented in _winapi.  Also, if it supports `CopyFileEx` instead, callers will be have access to flags such as `COPY_FILE_FAIL_IF_EXISTS` (not necessary here), `COPY_FILE_NO_BUFFERING` (may improve performance if the source file isn't in the system cache and is very large relative to the system cache), `COPY_FILE_COPY_SYMLINK` (not necessary here). Exposing these optional flags may be useful in the future, though not immediately useful for this patch. `_winapi.CopyFileEx` would only have to support `lpExistingFileName`, `lpNewFileName`, and `dwCopyFlags`. The other parameters (`lpProgressRoutine`, `lpData`, and `pbCancel`) can be included in the signature, but raise `NotImplementedError` if not `None`. Projects can depend on PyWin32 or ctypes if they need the full feature set.

I don't understand why `CopyFile` is getting labelled as "zero copy" for Windows. In a debugger you can follow its execution up to a synchronous I/O loop in `kernelbase!BaseCopyStream`, which basically consists of `ReadFile` and `WriteFile` with a fixed chunk size such as 1 MiB. The primary benefit in the common case is that files are opened with `FILE_FLAG_SEQUENTIAL_SCAN`, which optimizes the cache manager's read-ahead strategy for sequential access. We can get this already via `os.open` with `os.O_SEQUENTIAL`. However, `CopyFile[Ex]` also handles other scenarios, such optimizing the chunk size for remote files and [Offloaded Data Transfer](https://docs.microsoft.com/en-us/windows-hardware/drivers/storage/offloaded-data-transfer) (ODX), and it also does more than copy the default (anonymous) $DATA stream, as discussed below.

`CopyFile[Ex]` copies all $DATA streams in a file (e.g. the "Zone.Identifier" stream on downloaded files); timestamps (last access, last write, and change; but not creation); file attributes; extended file attributes; and security resource attributes (obscure; note this *is not* the owner, group, DACL and SACL from the file's security descriptor). Basically `CopyFile[Ex]` is `shutil.copyfile` +  `shutil.copystat`, so it can only be used in `shutil.copy2` (and thus `copytree` and `move`). 

This would expand the behavior of `copy2` on Windows to include copying extended attributes and alternate data streams. That in itself is reasonable since `copystat` copies extended attributes on other platforms. But it introduces an inconsistency when `copyfile` and `copystat` are called separately. To restore some consistency, `copyfile` could default to attempting to copy the named $DATA streams of a file, if supported by the target file system (i.e. NTFS). This would require a new built-in function that calls [`FindFirstStreamW`](https://msdn.microsoft.com/en-us/library/aa364424) and `FindNextStreamW` to list the named streams.

Regarding copying extended attributes in `copystat`, I think it's best to pretend they don't exist. There's no common use for them in Windows, so they're only accessible in the NT API (e.g. `NtQueryInformationFile`: `FileEaInformation`, `NtQueryEaFile`, `NtSetEaFile`), and they can't be set on reparse points (i.e symlinks, junctions, etc). I think the Linux subsystem (WSL) uses them in its per-user file system, such as for storing Linux file permissions in NTFS files, but Windows applications aren't supposed to touch those files.

giampaolo replied:
Thanks a lot for the detailed info! I appreciate it. I moved the syscall which now lives under `_winapi.CopyFileExW`. Good catch about the copying of metadata. That makes me realize that we don't have unit tests on Windows (or probably in general) which assert that no metadata is copied when using `copyfile`. 

It's a shame we cannot use `CopyFileEx` in `copyfile` though. I wonder if it would make sense to do the following in `copyfile`:

- create an empty dst file and collect its metadata 
- use CopyFileEx to copy src to dst
- reset dst file metadata before returning

What do you think?

giampaolo replied:
OK, I updated code which now uses `CopyFileEx` for `copy2()` only and updated the doc making it clear what happens on every platform. As for `FindFirstStreamW` it looks it complicates things quite a bit. I'll keep digging into it tomorrow and any reference would help of course. There are additional possibities such as using `fcopyfile` to update file content + metadata in one shot on OSX. I'll see if it makes sense to include them in this PR or not (probably not).

eryksun replied:
If `shutil.copyfile` were extended on Windows to copy all data streams in a file, then it wouldn't have to delete the named streams from the copy if implemented via `CopyFileEx`. Thus a new `os.liststreams` (i.e. `FindFirstStreamW`, `FindNextStreamW`) function wouldn't be required -- though adding it anyway would be nice to have. That said, undoing the copied meta-data starts to get fairly complicated when considering extended attributes and security resource attributes. It may be simpler for `copyfile` to use `os.open` with `os.O_SEQUENTIAL` -- in case you want to try benchmarking it. I still prefer using `CopyFileEx` in `shutil.copy2` (e.g. to get ODX with storage arrays that support it).

Assuming `copyfile` calls `CopyFileEx`, it would have to reset the file attributes to `FILE_ATTRIBUTE_ARCHIVE` via `_winapi.SetFileAttributes` (new), so as to not copy the source file attributes, such as read-only, hidden, system, temporary, and not-content-indexed. It would also need to set the last access and last write timestamps to match the creation time, which requires `os.stat` and `os.utime`.  Note that previously I stated that the last access timestamp gets copied as well, but on further inspection, the `NtSetInformationFile` call that sets the basic information of the target file actually zeros both the `CreationTime` and `LastAccessTime` fields, so only the `LastWriteTime` (data) and `ChangeTime` (meta-data) [1] timestamps are copied.

That leaves the problem of removing the copied extended attributes and security resource attributes. It's easy to *copy* extended attributes via `CreateFile` (from `hTemplateFile`) and extended attributes and streams via `CopyFileEx` or `CreateDirectoryEx` (from `lpTemplateDirectory`). (Directories can have data streams, e.g. the storage service creates zero-length streams named "Win32App_1" on application installation directories.) But, as far as I know, the Windows API doesn't provide a way to *remove* extended attributes. That's only possible in the native NT API. I suppose `copyfile` could ignore the fact that extended attributes are copied, if a practical, though not squeaky-clean, solution suffices.

Copying security resource attributes affects enterprise IT, where this kind of dynamic access control is used with central access policies. Like mandatory integrity label ACEs, resource-attribute ACEs are stored in the SACL (system access control list) of a file's security descriptor. Also, like mandatory integrity labels, setting them apparently doesn't require SeSecurityPrivilege. Below I set a "spam" string attribute with value "eggs" (i.e. `(RA;;;;;WD;("spam",TS,0,"eggs")`) on the source file, copy it via `CopyFileEx`, and then inspect the SACL of the destination file to confirm that the resource attribute was copied.

    >>> from win32file import *
    >>> from win32security import *
    >>> ATTRIBUTE_SECURITY_INFORMATION = 0x20

    >>> sd = ConvertStringSecurityDescriptorToSecurityDescriptor('S:(RA;;;;;WD;("spam",TS,0,"eggs"))', SDDL_REVISION_1)
    >>> sacl = sd.GetSecurityDescriptorSacl()
    >>> SetNamedSecurityInfo('test1.txt', SE_FILE_OBJECT, ATTRIBUTE_SECURITY_INFORMATION, None, None, None, sacl)

    >>> CopyFileEx('test1.txt', 'test2.txt')

    >>> sd2 = GetNamedSecurityInfo('test2.txt', SE_FILE_OBJECT, ATTRIBUTE_SECURITY_INFORMATION)
    >>> ConvertSecurityDescriptorToStringSecurityDescriptor(sd2, SDDL_REVISION_1, ATTRIBUTE_SECURITY_INFORMATION)
    'S:AI(RA;;;;;WD;("spam",TS,0x0,"eggs"))'

I still prefer to use `CopyFileEx` directly in `shutil.copy2`. If copying the last access time matters, it also needs `copystat`, but I don't see why it would really matter. Also, to be more consistent `copytree` should create directories via `_winapi.CreateDirectoryEx` (new) in order to copy extended attributes and named data streams (e.g. the "Win32App_1" streams mentioned above). It would split off the leaf node from `dst`. Call `os.makedirs(dst_parent)` to create the parent tree, if any, and call `_winapi.CreateDirectoryEx(src, dst, None)` to create the `dst` directory. However, this doesn't copy resource attributes from the source directory, so to be completely consistent it would have to copy them manually via `GetNamedSecurityInfo` and  `SetNamedSecurityInfo` for the `ATTRIBUTE_SECURITY_INFORMATION`. 

> [1] Windows Python doesn't use the meta-data change timestamp. Indirectly, this is because it's only implemented in NTFS/ReFS and is always 0 in FAT file systems. Directly, it's because `GetFileInformationByHandle` and `FindFirstFile` don't return it. (System calls provide the change time, but older API functions ignore it.) However, I think using `st_ctime` for the creation time on Windows is a wart in terms of cross-platform compatibility. I'd prefer using `st_birthtime` for the creation time, as on other platforms, and possibly changing `os.stat` and related functions to query the meta-data change time via `GetFileInformationByHandleEx`. Then add `_winapi.SetFileInformationByHandle` with support for `FileBasicInfo` to set all 4 timestamps and the file attributes in a single call.

giampaolo replied:
For the record: I tried using `os.O_SEQUENTIAL` on Windows for both src and dst fds and it makes copy a little slower.

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -74,6 +76,10 @@ Directory and files operations
+   .. versionchanged:: 3.8
+      Platform-specific zero-copy syscalls are used internally in order to copy
+      the file more efficiently. See
+      `platform-dependent efficient copy operations`_ section.
```
Since we're in Sphinx rather than plain `docutils`, this can be :ref:`shutil-platform-dependent-efficient-copy-operations`.

I'd also be inclined to shorten the initial sentence even further. Something like "Optimised platform-specific calls are used to copy files more efficiently. See ... for more details."

giampaolo replied:
Done, thanks for the tip.

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -370,6 +395,26 @@ Directory and files operations
+
+Starting from Python 3.8 :func:`copyfile` uses platform-specific "zero-copy"
+syscalls such as :func:`os.sendfile` in order to copy the file more efficiently
+(see `bpo-33671 <https://bugs.python.org/issue33671>`_).
```
There's a shorthand available for BPO links:

    :issue:`33671`

giampaolo replied:
Done.

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -370,6 +395,26 @@ Directory and files operations
+All functions relying on :func:`copyfile` will benefit from the same speedup.
+These are :func:`shutil.copy`, :func:`shutil.copy2`, :func:`shutil.copytree`
+and :func:`shutil.move`.
+
```
Given @eryksun's comments about how the relevant Windows API works, it may be better to just name the platform specific APIs here, without going into details on exactly how those APIs achieve their performance gains relative to a Python level copy operation.

giampaolo replied:
Agreed. Done.

ncoghlan commented on Doc/library/shutil.rst:
```diff
@@ -657,3 +702,5 @@ Querying the size of the output terminal
 .. _`Other Environment Variables`:
    http://pubs.opengroup.org/onlinepubs/7908799/xbd/envvar.html#tag_002_003
 
+.. _`platform-dependent efficient copy operations`:
```
Given a switch to `ref` syntax above, you'll be able to drop these two lines.

ncoghlan commented on Doc/whatsnew/3.8.rst:
```diff
@@ -90,10 +90,20 @@ New Modules
+  "zero-copy" syscalls such as :func:`os.sendfile` in order to copy the file
+  more efficiently.
+  "zero-copy" means that the copying operation occurs within the kernel,
+  avoiding the use of userspace buffers as in "``outfd.write(infd.read())``".
```
As with the main docs, glossing over the details is likely a better option than trying to come up with a cross-platform summary of the semantics of kernel level copy operations.

giampaolo replied:
I opted for:

> "fast-copy" means that the copying operation occurs within the kernel, avoiding
> the use of userspace buffers in Python as in "``outfd.write(infd.read())``".

ncoghlan reviewed (commented):
This is starting to look very nice to me :)

@eryksun raises some valid points regarding the Windows side of the patch (in particular, moving the `CopyFile` wrapper out into `_winapi`), and I have one remaining technical comment (since the `_HAVE_*API*` flags can never become True dynamically once they're false, I'd like to ensure we end up with `_copyfileobj2` being a simple alias for `copyfileobj` in those cases).

There are also some remaining cosmetic comments regarding the value of being vague about exactly how the kernel APIs are achieving their performance improvements relative to the user-space copy operations.

giampaolo committed:
Merge branch 'master' into shutil-zero-copy

giampaolo committed:
rename function using 'fastcopy' prefix instead of 'zerocopy'

giampaolo committed:
use :ref: in rst doc

giampaolo committed:
change wording in doc

giampaolo committed:
add test to make sure sendfile() doesn't get called aymore in case it doesn't support file to file copies

giampaolo committed:
move CopyFileW in _winapi and actually expose CopyFileExW instead

giampaolo committed:
fix line endings

eryksun commented on Lib/shutil.py:
```diff
@@ -117,9 +228,13 @@ def copyfile(src, dst, *, follow_symlinks=True):
     if not follow_symlinks and os.path.islink(src):
         os.symlink(os.readlink(src), dst)
     else:
+        if os.name == 'nt':
```
This cannot be used in `shutil.copyfile`. My previous comment explains why in detail. It can be used in `shutil.copy2`, but this makes separate `copyfile` and `copystat` calls inconsistent with `copy2`. I suggested a possible way to address this.

giampaolo committed:
add tests for mode bits

giampaolo committed:
add docstring

giampaolo committed:
remove test file mode class; let's keep it for later when Istart addressing OSX fcopyfile() specific copies

giampaolo committed:
update doc to reflect new changes

giampaolo committed:
update doc

ncoghlan reviewed (approved):
(Approving for docs and systems that use the sendfile path)

giampaolo committed:
adjust tests on win

giampaolo committed:
fix argument clinic error

giampaolo committed:
update doc

giampaolo committed:
OSX: expose copyfile(3) instead of fcopyfile(3); also expose flags arg to python

giampaolo committed:
osx / copyfile: use path_t instead of char

giampaolo committed:
do not set dst name in the OSError exception in order to remain consistent with platforms which cannot do that (e.g. linux)

giampaolo committed:
add same file test

giampaolo committed:
add test for same file

giampaolo committed:
have osx copyfile() pre-emptively check if src and dst are the same, otherwise it will return immedialtey and src file content gets deleted

giampaolo committed:
turn PermissionError into appropriate SameFileError

giampaolo committed:
expose ERROR_SHARING_VIOLATION in order to raise more appropriate SameFileError

giampaolo committed:
honour follow_symlinks arg when using CopyFileEx

giampaolo committed:
update Misc/NEWS

Ryanb58 commented on Doc/library/shutil.rst:
```diff
@@ -370,6 +395,28 @@ Directory and files operations
+
+Starting from Python 3.8 all functions involving a file copy (:func:`copyfile`,
+:func:`copy`, :func:`copy2`, :func:`copytree`, and :func:`move`) use
+platform-specific "fast-copy" syscalls in order to copy the file more
```
Is there a difference between `fast-copy` and `zero-copy` sys calls? #learning

giampaolo replied:
Initially I phrased this as "zero-copy" because that's what `os.sendfile()` actually does under the hoods. Then it turns out Windows does something different (see https://github.com/python/cpython/pull/7160#discussion_r191640338). As such it made more sense to stick with the generic "fast-copy" term.  =)

Ryanb58 commented on Lib/test/test_shutil.py:
```diff
@@ -1829,6 +1891,246 @@ def test_move_dir_caseinsensitive(self):
             os.rmdir(dst_dir)
 
+
+class _ZeroCopyFileTest(object):
```
Do the cpython unit tests get run on the different platforms you have defined in these unittests?

Ryanb58 reviewed (approved):
LGTM

Awesome job!

giampaolo committed:
expose CreateDirectoryEx mock

giampaolo committed:
change C type

giampaolo committed:
CreateDirectoryExW actual implementation

giampaolo committed:
provide specific makedirs() implementation for win

giampaolo committed:
Merge branch 'shutil-zero-copy-8' of https://github.com/giampaolo/cpython into shutil-zero-copy-8

giampaolo committed:
fix typo

giampaolo committed:
skeleton for SetNamedSecurityInfo

giampaolo committed:
get security info for src path

giampaolo committed:
finally set security attrs

giampaolo committed:
add unit tests

giampaolo committed:
mimick os.makedirs() behavior and raise if dst dir exists

giampaolo committed:
set 2 paths for OSError object

giampaolo committed:
set 2 paths for OSError object

giampaolo committed:
expand windows test

giampaolo committed:
in case of exception on os.sendfile() set filename and filename2 exception attributes

giampaolo committed:
set 2 filenames (src, dst) for OSError in case copyfile() fails on OSX

giampaolo committed:
update doc

giampaolo committed:
do not use CreateDirectoryEx() in copytree() if source dir is a symlink (breaks test_copytree_symlink_dir); instead just create a plain dir and remain consistent with POSIX implementation

giampaolo committed:
use bytearray() and readinto()

giampaolo committed:
use memoryview() with bytearray()

giampaolo committed:
refactoring + introduce a new _fastcopy_binfileobj() fun

giampaolo committed:
remove CopyFileEx and other C wrappers

giampaolo committed:
remove code related to CopyFileEx

giampaolo committed:
Recognize binary files in copyfileobj()
...and use fastest _fastcopy_binfileobj() when possible

giampaolo committed:
set 1MB copy bufsize on win; also add a global _COPY_BUFSIZE variable

giampaolo committed:
use ctx manager for memoryview()

giampaolo committed:
update doc

giampaolo committed:
remove outdated doc

giampaolo committed:
remove last CopyFileEx remnants

giampaolo committed:
OSX - use fcopyfile(3) instead of copyfile(3)

...as an extra safety measure: in case src/dst are "exotic" files (non
regular or living on a network fs etc.) we better fail on open() instead
of copyfile(3) as we're not quite sure what's gonna happen in that
case.

giampaolo committed:
update doc

eryksun commented on Lib/shutil.py:
```diff
@@ -72,14 +83,124 @@ class RegistryError(Exception):
+            if not n:
+                break
+            elif n < length:
+                fdst_write(mv[:n])
```
In my test case I used another `with` block to release this `mv[:n]` view, rather than depend on implicit deallocation to release it. For example:

    >>> b = bytearray(100)
    >>> mv1 = memoryview(b)
    >>> mv2 = mv1[:10]
    >>> mv1.release()
    >>> b.append(0)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    BufferError: Existing exports of data: object cannot be re-sized
    >>> mv2.release()
    >>> b.append(0)
    >>> len(b)
    101

This `bytearray` is internal, but is there any issue with memory usage in garbage-collected versions of Python (e.g. Jython, IronPython) if the views on the buffer (1 MiB  in Windows) aren't released explicitly? If not you can remove the first `with` block as well.

giampaolo replied:
Uhm... yes, given the big bufsize I think it makes sense to also immediately release the sliced memoryview.

ncoghlan replied:
@eryksun If I recall correctly, memory views inadvertently keeping large memory buffers alive on GC based implementations was a key driver in adding context management support to `memoryview` in the first place, so that's definitely a concern worth keeping in mind for this kind of code.

eryksun commented on Lib/shutil.py:
```diff
@@ -72,14 +83,124 @@ class RegistryError(Exception):
+
+def _is_binary_files_pair(fsrc, fdst):
+    return hasattr(fsrc, 'readinto') and \
+        isinstance(fsrc, io.BytesIO) or 'b' in getattr(fsrc, 'mode', '') and \
```
Which objects provide `readinto` in text mode? Is it worth the function call and extra tests rather than handling `AttributeError` (no `readinto`) and `TypeError` (can't write bytes) with an inline try-except in `copyfileobj`?

giampaolo replied:
I think catching `TypeError` on `fdst.write()` is too risky as we can deal with any kind of custom file-like object being passed here. It must be noted that the extra cost of this function is payed by users of `copyfileobj()` only (e.g. `tarfile` and `zipfile` modules). `copyfile()` (and others) will skip this check and call `_copybinfileobj()` directly.

giampaolo replied:
I benchmarked `_copybinfileobj()` (I hadn't yet) and it turns out it's only slightly faster for 512MB files but considerably slower for 8MB and 128KB files so I am gonna remove it.

eryksun replied:
I don't know how you're testing, but the performance difference with `readinto` depends on whether the source file is already in the system cache. Otherwise, of course memory operations will be dwarfed by considerably slower disk I/O.

giampaolo replied:
This is how I'm testing it:

```bash
$ python -c "import os; f = open('f1', 'wb'); f.write(os.urandom(8 * 1024 * 1024))"
$ time ./python -m timeit -s 'import shutil; p1 = "f1"; p2 = "f2"' 'shutil.copyfile(p1, p2)'
```

giampaolo replied:
Sorry, wrong cmdline. This is the correct one:

```bash
$ python -c "import os; f = open('f1', 'wb'); f.write(os.urandom(8 * 1024 * 1024))"
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
```

Original:

```
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
200000 loops, best of 5: 1.36 usec per loop
```
Memoryview / readinto:
```
$ ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"
100000 loops, best of 5: 3.22 usec per loop
```

eryksun replied:
Does the original case use a 1 MiB `length` for the 8 MiB file, or is that the original 16 KiB length?

giampaolo replied:
Both cases use 1MiB.

eryksun replied:
Originally I tested 128 MiB source files, which appears to be a sweet spot in my case. There's a setup overhead cost that makes using `readinto` more expensive if the source file is less than about 64 MiB. At 128 MiB I observe about 30% improvement with `readinto`. At 256 MiB and 512 MiB, it's only about a 10% improvement (observed varying between 7% and 12%).

eryksun replied:
I checked Linux with the default 16 KiB `length`. Using `readinto` provides no benefit on larger files and degrades performance on smaller files. Given only a modest 10% improvement in Windows with large files, I think your decision to remove it is the right call. It's not worth the bookkeeping overhead to keep two versions, and it's not clear even on Windows where a good dividing line would be if we `[f]stat` the source file to check `st_size`.

giampaolo replied:
I wrote a batch script to figure out timings on Windows more easily and this is the result (first value is original `copyfileobj()` implementation, second value is the `memoryview()` variant):

```
8MB file
1000 loops, best of 5: 343 usec per loop
500 loops, best of 5: 478 usec per loop

64MB file
500 loops, best of 5: 474 usec per loop
500 loops, best of 5: 554 usec per loop

128MB file
200 loops, best of 5: 1.06 msec per loop
500 loops, best of 5: 640 usec per loop

256MB file
1 loop, best of 5: 286 msec per loop
5 loops, best of 5: 36.1 msec per loop

512MB file
1 loop, best of 5: 293 msec per loop
5 loops, best of 5: 36.7 msec per loop
```

I think the `memoryview()` variant after 128MB is so much faster that it is worth to have the dual implementation and use it **from `copyfile()` function only** if on Windows and file size > 128MB. I will do it in my other PR/branch.

On the other hand, the same test on Linux shows there is no relevant difference for 512 MB files and a performance degradation for smaller ones.

giampaolo replied:
Done: https://github.com/python/cpython/pull/7681/files

eryksun replied:
This command is wrong:

    ./python  -m timeit -s "import shutil; f1 = open('f1', 'rb'); f2 = open('f2', 'wb')" "shutil.copyfileobj(f1, f2)"

The files need to be opened for each pass of the loop, not the setup. That explains the unexpected results. I corrected it to open the files in the loop statement instead of the setup and tested a broad range of file sizes. In the table below all times are best of five for the give number of loops, and normalized overall to make the 64KiB result in the RI_S column equal to 100 time units. I discuss the RI_S case in more detail below.

```
   SIZE | LOOPS |   R_16K ||    R_1M   %CHNG |  RI_1M   %CHNG |  RI_S   %CHNG
--------+-------+---------||-----------------+----------------+--------------
  1 GiB |     5 | 1060870 || 1003478    -5.4 | 977391    -7.9 | 963478   -9.2
512 MiB |    10 |  323478 ||  283478   -12.4 | 217391   -32.8 | 213913  -33.9
256 MiB |    20 |  163304 ||  146435   -10.3 | 112870   -30.9 | 110957  -32.1
128 MiB |    40 |   80174 ||   74609    -6.9 |  55478   -30.8 |  55478  -30.8
 64 MiB |    80 |   39652 ||   34783   -12.3 |  27304   -31.1 |  27130  -31.6
 32 MiB |   160 |   19478 ||   17913    -8.0 |  13809   -29.1 |  13478  -30.8
 16 MiB |   320 |    9739 ||    8887    -8.7 |   6887   -29.3 |   6835  -29.8
  8 MiB |   640 |    4904 ||    4713    -3.9 |   3652   -25.5 |   3548  -27.7
  4 MiB |  1280 |    2504 ||    2348    -6.2 |   1948   -22.2 |   1913  -23.6
  2 MiB |  2560 |    1150 ||    1193     3.7 |   1002   -12.9 |    991  -13.8
  1 MiB |  5120 |     649 ||     697     7.4 |    659     1.5 |    663    2.2
512 KiB | 10240 |     388 ||     553    42.5 |    499    28.6 |    322  -17.0
256 KiB | 20480 |     245 ||     459    87.3 |    410    67.3 |    205  -16.3
128 KiB | 30720 |     170 ||     388   128.2 |    362   112.9 |    139  -18.2
 64 KiB | 40960 |     123 ||     353   187.0 |    343   178.9 |    100  -18.7

    R_16K -- read 16 KiB
    R_1M  -- read 1 MiB
    RI_1M -- readinto 1 MiB
    RI_S  -- readinto source size up to 1 MiB
```

Originally I had tested at 128 MiB with a custom test script to focus on the effects of cached vs non-cached I/O. I assumed the results would be similar for other cases. As shown in the RI_1M column, that's basically true for files larger than 1 megabyte. But there's a significant performance degradation for smaller files.  In the RI_S case, I address this by calling `os.fstat` on the source file to cap the `length` of the `bytearray` at its size. This avoids wastefully over-allocating a zeroed `byterray`. 

RI_S also experiments with calling `SetInformationByHandle` : `FileEndOfFileInfo` (prototyped using ctypes) to avoid having to repeatedly extend the file when `length` is less than the size of the source file. (Note that Python's `truncate` method is of no use here since it zeros the file.) `CopyFileEx` does this, so I figured it was worth a try. This does provide a modest performance increase. (Note that I mistakenly included the `length == size` boundary, which is apparent in the 1 MiB trial.) I don't know if it's significant enough to justify implementing `_winapi.SetFileInformationByHandle`. 

If I have time, I may run another experiment using `mmap` to read into a sliding window of the destination file. It already implements setting the end of the file, albeit with the less efficient combination of `SetFilePointer` and `SetEndOfFile`. (This way requires 4 system calls instead of 1 to set the end of the file.)

Below is the code for RI_S:

     import os

    _WINDOWS = (os.name == 'nt')

    if _WINDOWS:
        import msvcrt
        import ctypes
        from ctypes import wintypes
        
        COPY_BUFSIZE = 1024 *1024
        
        kernel32 = ctypes.WinDLL('kernel32', use_last_error=True)

        FileEndOfFileInfo = 6

        FILE_INFO_BY_HANDLE_CLASS = wintypes.DWORD
        kernel32.SetFileInformationByHandle.argtypes = (
            wintypes.HANDLE,           # _In_ hFile
            FILE_INFO_BY_HANDLE_CLASS, # _In_ FileInformationClass
            wintypes.LPVOID,           # _In_ lpFileInformation
            wintypes.DWORD)            # _In_ dwBufferSize

    def copyfileobj(fsrc, fdst, length=COPY_BUFSIZE):
        size = os.fstat(fsrc.fileno()).st_size
        if length > size:
            length = size
        elif _WINDOWS:
            info = wintypes.LARGE_INTEGER(size)
            if not kernel32.SetFileInformationByHandle(
                        msvcrt.get_osfhandle(fdst.fileno()), FileEndOfFileInfo, 
                        ctypes.byref(info), ctypes.sizeof(info)):
                raise ctypes.WinError(ctypes.get_last_error())
        fsrc_readinto = fsrc.readinto
        fdst_write = fdst.write
        with memoryview(bytearray(length)) as mv:
            while True:
                n = fsrc_readinto(mv)
                if not n:
                    break
                elif n < length:
                    with mv[:n] as smv:
                        fdst_write(smv)
                else:
                    fdst_write(mv)

giampaolo replied:
Interesting. Thanks for the very detailed benchmark. I updated the other branch which now dynamically sets `memoryview()` size based on file size (https://github.com/python/cpython/pull/7681/commits/93ebc1ffdc547f4057a164922c7962186ef1c320) and I confirm using the `readinto()` variant is faster also for smaller files.

giampaolo replied:
https://github.com/python/cpython/pull/7681 was merged
